src/main/java/com/artipie/asto/cache/FromRemoteCache.java

```diff
@@ -51,18 +51,20 @@ public FromRemoteCache(final Storage storage) {
     }
 
     @Override
-    public CompletionStage<? extends Content> load(
-        final Key key, final AsyncContent remote, final CacheControl control
+    public CompletionStage<Optional<? extends Content>> load(
+        final Key key, final Remote remote, final CacheControl control
     ) {
         return remote.get().handle(
             (content, throwable) -> {
-                final CompletionStage<? extends Content> res;
-                if (throwable == null) {
-                    res = this.storage.save(key,  new Content.From(content.size(), content))
-                        .thenCompose(nothing -> this.storage.value(key));
+                final CompletionStage<Optional<? extends Content>> res;
+                if (throwable == null && content.isPresent()) {
+                    res = this.storage.save(
+                        key,  new Content.From(content.get().size(), content.get())
+                    ).thenCompose(nothing -> this.storage.value(key))
+                        .thenApply(Optional::of);
                 } else {
                     res = new FromStorageCache(this.storage)
-                        .load(key, new AsyncContent.Failed(throwable), control);
+                        .load(key, new Remote.Failed(throwable), control);
```

Comment:
@olenagerasimova we are going to this code branch when `content.isPresent() == false` and then `throwable` is `null`. Case when `throwable` is `null` and content is missing requires special handling. Also this case is not covered with tests